misc/futures-bounded/src/bounded.rs

```diff
@@ -0,0 +1,186 @@
+use std::future::Future;
+use std::task::{Context, Poll, Waker};
+use std::time::Duration;
+
+use futures_timer::Delay;
+use futures_util::future::{select, BoxFuture, Either};
+use futures_util::stream::FuturesUnordered;
+use futures_util::{ready, FutureExt, StreamExt};
+
+use crate::{PushResult, Timeout};
+
+/// Represents a set of (Worker)-[Future]s.
+///
+/// This wraps [FuturesUnordered] but bounds it by time and size.
+/// In other words, each worker must finish within the specified time and the set never outgrows its capacity.
+pub struct BoundedWorkers<O> {
+    timeout: Duration,
+    capacity: usize,
+    inner: FuturesUnordered<BoxFuture<'static, Result<O, Timeout>>>,
+
+    empty_waker: Option<Waker>,
+    full_waker: Option<Waker>,
+}
+
+impl<O> BoundedWorkers<O> {
+    pub fn new(timeout: Duration, capacity: usize) -> Self {
+        Self {
+            timeout,
+            capacity,
+            inner: Default::default(),
+            empty_waker: None,
+            full_waker: None,
+        }
+    }
+}
+
+impl<O> BoundedWorkers<O> {
+    /// Push a worker into the set.
+    /// This method adds the given worker to the set and returns [PushResult::Ok].
+    /// If length of the set is equal the capacity, this method returns [PushResult::BeyondCapacity]
+    /// and worker is ignored.
+    pub fn try_push<F>(&mut self, worker: F) -> PushResult
```

Comment:
Ideally, this API doesn't use `PushResult` because it will never return the 2nd error regarding IDs. We shouldn't make users handle errors that will never occur :)